[98091996] Deploy multiple apps to tsuru to load test app deployment.

dhilton opened the pull request:
https://www.pivotaltracker.com/story/show/98091996

**What**

This PR sets up a load test for creating and deploying apps to a Tsuru platform endpoint. We make use of the SSH helper created as part of the integration tests, to enable us to deploy using `git push` as part of the deployment load test. We have used https://github.com/flood-io/ruby-jmeter as a DSL to speed up the writing of tests. However, in this usecase we have an issue with calling an os process sampler (a way of running a command outside of the load test). 

**How to test**

Clone the repo and then follow the README.md setup instructions (as we need to test those as part of the PR) then run a set of deploy apps and cleanup apps scripts. You should get a set of results from jmeter to prove that the tests have been run and output from running `tsuru app-list` will show you the created apps after running the `deploy_apps.rb` generated jmeter run. 

**Who can test**

Anyone on the core team apart from @saliceti

dhilton committed:
Updated README on how to install and current limitations

dhilton committed:
Adding gemfile, .gitignore to setup the project in a correct manner.

dhilton committed:
Performance tests that allow you to loadtest app deployment.

A test and cleanup script to generate a jMeter jmx file to run a load test of creating an app and
deploying an app via `git push` to a gandalf server.

There is an immediate step where command args and environment variables need to be added in jmeter,
this has been documented in the README.md

We await an up stream fix to remove this workaround: https://github.com/flood-io/ruby-jmeter/issues/43

dhilton committed:
Added example jmx file to illustrate the issue with command args passing to process os sampler.

This gives context to the workaround documented in the README.md

dcarley commented on dsl/cleanup_apps.rb:
```diff
@@ -0,0 +1,45 @@
+  threads count: ARGV[1].to_i, loops: ARGV[2].to_i do
+    Once do
+      post url: auth_url,
+        raw_body: '{ "password": "admin123" }' do
```
We shouldn't store the username and password in plaintext or as fixed strings. Even if they aren't secure at the moment, they will be at some point in the near future and this will break. Is it feasible to extract them from the vault, like we [do for the end-to-end tests](https://github.com/alphagov/tsuru-ansible/blob/463fae35f6862d15d331185bd957a3c42502354c/Makefile#L72-L80)?

keymon replied:
In the end-to-end test we are lucky of being in the same repo than the vault. I think we need common secret manager, as we grow in number of repos and components.

That generates a .xml file, which we should not commit as it would contain the secret credentials. Maybe it is worth it to ignore .xml (I think it is the case)

dcarley commented on dsl/git_push_test.jmx:
```diff
@@ -0,0 +1,254 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<jmeterTestPlan version="1.2" properties="2.8" jmeter="2.13 r1665067">
```
Was this file meant to be committed? The README doesn't mention how to use it and suggests that something similar to this will be generated for you.

dhilton replied:
This commit mentioned it: https://github.com/alphagov/tsuru-performance-tests/commit/913fe39efe89b25811eda82237ddc458ec2fa789  - happy to remove it if you think it's overkill.

dcarley replied:
Sorry, didn't see that. I'm no clearer about how to read it though. Is it these sections that have been added?

```
+          <elementProp name="SystemSampler.arguments" elementType="Arguments" guiclass="ArgumentsPanel" testclass="Arguments" testname="OsProcessSampler" enabled="true">
+            <collectionProp name="Arguments.arguments">
+              <elementProp name="" elementType="Argument">
+                <stringProp name="Argument.name"></stringProp>
+                <stringProp name="Argument.value">push</stringProp>
+                <stringProp name="Argument.metadata">=</stringProp>
+              </elementProp>
+              <elementProp name="" elementType="Argument">
+                <stringProp name="Argument.name"></stringProp>
+                <stringProp name="Argument.value">git@ci-gandalf.tsuru2.paas.alphagov.co.uk:davas-${app_id}.git</stringProp>
+                <stringProp name="Argument.metadata">=</stringProp>
+              </elementProp>
+              <elementProp name="" elementType="Argument">
+                <stringProp name="Argument.name"></stringProp>
+                <stringProp name="Argument.value">master</stringProp>
+                <stringProp name="Argument.metadata">=</stringProp>
+              </elementProp>
+            </collectionProp>
+          </elementProp>
```

```
+          <elementProp name="SystemSampler.environment" elementType="Arguments" guiclass="ArgumentsPanel" testclass="Arguments" testname="OsProcessSampler" enabled="true">
+            <collectionProp name="Arguments.arguments">
+              <elementProp name="HOME" elementType="Argument">
+                <stringProp name="Argument.name">HOME</stringProp>
+                <stringProp name="Argument.value">/tmp/tsuru_tmp</stringProp>
+                <stringProp name="Argument.metadata">=</stringProp>
+              </elementProp>
+              <elementProp name="GIT_SSH" elementType="Argument">
+                <stringProp name="Argument.name">GIT_SSH</stringProp>
+                <stringProp name="Argument.value">/tmp/tsuru_tmp/ssh-wrapper</stringProp>
+                <stringProp name="Argument.metadata">=</stringProp>
+              </elementProp>
+            </collectionProp>
+          </elementProp>
```

Maybe list that in the README? I wouldn't have understood how to get from those bullet points to that XML.

dhilton replied:
The README.md documents how to add those variables in the Jmeter GUI - it's only for extra reference really, how about I just extract those relevant parts of the xml and put in the README.md for reference? It's just in case someone has some weird issues with the variables being miss set by the gui.

dcarley replied:
Oh, I see. Hadn't got as far as running `jmeter` yet. If I'm not expected to spelunk in XML then I think we can omit this file.

dcarley commented on README.md:
````diff
@@ -1,3 +1,49 @@
+To run the tests:
+
+```
+ruby deploy_apps.rb ruby deploy_apps.rb <TSURU_TARGET_URL> 100 100
````
This command should be prefixed with `bundle exec` to use the dependencies installed with `bundle install` in the commands above.

dcarley replied:
Oh, I've just realised the dependencies aren't listed in `Gemfile`, but they should be because I get this error:

```
➜  dsl git:(98091996-deploy-mulitple-apps-tsuru) bundle exec ruby deploy_apps.rb https://dcarley-api.tsuru.paas.alphagov.co.uk/ 100 100
deploy_apps.rb:2:in `require': cannot load such file -- ruby-jmeter (LoadError)
        from deploy_apps.rb:2:in `<main>'
```

dcarley replied:
It also says `ruby deploy_apps.rb` twice. Thought that was my typo at first ;)

dcarley commented on README.md:
````diff
@@ -1,3 +1,49 @@
+To clean up afterwards run:
+
+```
+ruby cleanup_apps.rb <TSURU_TARGET_URL> 100 100
````
Ditto ^ about `bundle exec`

dcarley commented on README.md:
```diff
@@ -1,3 +1,49 @@
+Brew install jmeter
+git clone https://github.com/alphagov/tsuru-performance-testing
+cd ./tsuru-performance-testing
+bundle install
```
It looks like `deploy_apps.rb` and `cleanup_apps.rb` need to be run from the `dsl/` directory. Should we include a `cd dsl` in these instructions?

dcarley commented on Gemfile:
```diff
@@ -0,0 +1 @@
+"rubocop"
```
I think this needs to be `gem "rubocop"` and have a `source "https://rubygems.org"` line. Otherwise `bundle install` doesn't do anything:

```
➜  tsuru-performance-tests git:(98091996-deploy-mulitple-apps-tsuru) ✗ bundle install
The Gemfile specifies no dependencies
```

dcarley commented on dsl/deploy_apps.rb:
```diff
@@ -0,0 +1,89 @@
+  puts '\n\n------- Tsuru perf test: ERROR! --------'
+  puts '      Usage:'
+  puts '      Please specify a tsuru target domain, a thread'
+  puts '  count and a loop count for each thread.\n '
```
Is this usage message and the command in the README correct? It doesn't seem to match the variable assignments of `environment` and `host_suffix` on lines 18-21.

dcarley commented on dsl/deploy_apps.rb:
```diff
@@ -0,0 +1,89 @@
+
+    os_process_sampler 'SystemSampler.command'     => 'git',
+                       'SystemSampler.arguments'   => %w(push origin master),
+                       'SystemSampler.directory'   => '#{tsuru_home}/example-java-jetty',
```
Should this be in double quotes? I get this line in my `.jmx` file, which looks like it should have been expanded:

```
          <stringProp name="SystemSampler.directory">#{tsuru_home}/example-java-jetty</stringProp>
```

dcarley commented on dsl/deploy_apps.rb:
```diff
@@ -0,0 +1,89 @@
+    end
+
+    post url: '/apps',
+         raw_body: '{"name":"davas-${app_id}","plan":{"name":""},"platform":"java","pool":"","teamOwner":""}'
```
What does "davas" mean? :hushed:

dhilton replied:
Removed reference to 'davas' replaced with testapp.

dhilton committed:
Fix for gemfile dependencies for Rubocup.

dhilton committed:
Changed name of test app

dhilton committed:
Fixed typo, changed ruby calls to use bundle exec to sandbox the runtime environment.

dhilton committed:
Remove example jmeter file as relevant XML is now in the README

dcarley commented on README.md:
```diff
@@ -1,3 +1,97 @@
+
+
+
+### Example OS Process Sampler
```
From meatspace: if we don't think the upstream bug is going to be fixed any time soon then it's probably easier for us to munge the resulting XML with Ruby. Although we'll still have to run Jmeter by hand.

keymon replied:
I created a shell script for this:

```
DEPLOY_ENV=hector070715

cat <<EOF > /tmp/system_sampler.txt
<SystemSampler guiclass="SystemSamplerGui" testclass="SystemSampler" testname="OsProcessSampler" enabled="true">
          <boolProp name="SystemSampler.checkReturnCode">false</boolProp>
          <stringProp name="SystemSampler.expectedReturnCode">0</stringProp>
          <stringProp name="SystemSampler.command">/usr/bin/git</stringProp>
          <elementProp name="SystemSampler.arguments" elementType="Arguments" guiclass="ArgumentsPanel" testclass="Arguments" testname="OsProcessSampler" enabled="true">
            <collectionProp name="Arguments.arguments">
              <elementProp name="" elementType="Argument">
                <stringProp name="Argument.name"></stringProp>
                <stringProp name="Argument.value">push</stringProp>
                <stringProp name="Argument.metadata">=</stringProp>
              </elementProp>
              <elementProp name="" elementType="Argument">
                <stringProp name="Argument.name"></stringProp>
                <stringProp name="Argument.value">git@${DEPLOY_ENV}-gandalf.tsuru2.paas.alphagov.co.uk:testapp-\${app_id}.git</stringProp>
                <stringProp name="Argument.metadata">=</stringProp>
              </elementProp>
              <elementProp name="" elementType="Argument">
                <stringProp name="Argument.name"></stringProp>
                <stringProp name="Argument.value">master</stringProp>
                <stringProp name="Argument.metadata">=</stringProp>
              </elementProp>
            </collectionProp>
          </elementProp>
          <elementProp name="SystemSampler.environment" elementType="Arguments" guiclass="ArgumentsPanel" testclass="Arguments" testname="OsProcessSampler" enabled="true">
            <collectionProp name="Arguments.arguments">
              <elementProp name="HOME" elementType="Argument">
                <stringProp name="Argument.name">HOME</stringProp>
                <stringProp name="Argument.value">/tmp/tsuru_tmp</stringProp>
                <stringProp name="Argument.metadata">=</stringProp>
              </elementProp>
              <elementProp name="GIT_SSH" elementType="Argument">
                <stringProp name="Argument.name">GIT_SSH</stringProp>
                <stringProp name="Argument.value">/tmp/tsuru_tmp/ssh-wrapper</stringProp>
                <stringProp name="Argument.metadata">=</stringProp>
              </elementProp>
            </collectionProp>
          </elementProp>
          <stringProp name="SystemSampler.directory">/tmp/tsuru_tmp/example-java-jetty</stringProp>
        </SystemSampler>
EOF

awk '/\<SystemSampler/ { system("cat /tmp/system_sampler.txt") }  /<SystemSampler/,/<\/SystemSampler>/ { next; }  // { print } ' < jmeter.jmx > jmeter2.jmx



```

dhilton committed:
Updated dependencies

dhilton committed:
Clean up of help messages and usage notes

Changed how the commandline tools are invoked, bringing a common interface
and error reporting between the two scripts.

keymon commented on README.md:
````diff
@@ -1,3 +1,97 @@
+bundle exec ruby deploy_apps.rb <TSURU_ENVIRONMENT_NAME>  <TSURU_TARGET_HOST> 100 100
+```
+
+This will generate a jmeter.jmx file which you can load into the Jmeter gui or use in a headless mode.
````
I think we should add how to run jmeter from the console. For instance, something like this:

This will generate a jmeter.jmx file which you can load into Jmeter in GUI mode:

```
jmeter -t jmeter2.jmx
```

Or in the headless mode with `-n` and saving results for later analysis `-l my_results.jtl`

```
jmeter -n -t jmeter2.jmx -l my_results.jtl
```

dhilton replied:
The ruby jmeter dsl offers a headless mode; currently we don't use it because of the os process sampler issue. This is a good alternative though to having to run the jmeter GUI - I've updated the README.md to include this option for the deploy_apps.rb script (as that is the only one that can't be run in the correct headless method).

keymon commented on README.md:
````diff
@@ -1,3 +1,97 @@
+bundle exec ruby cleanup_apps.rb <TSURU_ENVIRONMENT_NAME> <TSURU_TARGET_HOST> 100 100
+```
+
+
````
Can we add instructions of installing jmeter on brew and ubuntu, and a link to the tool. For instance:

# Requirements and dependencies

We use apache [jmeter](http://jmeter.apache.org/) to run the tests. To install it:
- On MacOSX with homebrew: `brew install jmeter`
- On Ubuntu: `apt-get install jmeter`
- Other approaches: http://jmeter.apache.org/usermanual/get-started.html

For generating the test templates, we use ruby 2.2.2. It is recommended the use of [rvm](https://rvm.io/) or [rbenv](http://rbenv.org/).

dhilton replied:
I've updated the readme with these instructions.

keymon commented:
I am able to generate the jmeter.jmx, and run the tests. But it seems a little bit inestable, I managed to break it, leaving a lot of applications created and locked. Now I am having doing clean up process, getting:

```
App locked by tsr, running POST /apps/testapp-1/repository/clone. Acquired in 2015-07-07T16:10:43Z
```

It seems the application got lock while doing the push at some point, probably while running and stopping the test. I had to delete the lock using this script:

```
for i in $(tsuru app-list | sed -n 's/.*\(testapp-[0-9]*\) .*/\1/p'); do echo $i; tsuru-admin app-unlock -a $i -y || break; done
```

keymon commented on dsl/cleanup_apps.rb:
```diff
@@ -0,0 +1,50 @@
+    { name: "Content-Type", value: "application/json" }
+  ]
+  auth_url = "/users/" + CGI.escape("administrator@gds.tsuru.gov") + "/tokens"
+  threads count: ARGV[1].to_i, loops: ARGV[2].to_i do
```
This should be: `threads count: thread_count, loops: loop_count do`. Currently it sets threads to 0 and loops to the value of threads :)

keymon committed:
Fix bug reading arguments in cleanup.rb

We were reading the wrong arguments. Copy the same behaviour than
`deploy_apps.rb`

dhilton committed:
Update README to contain more platform install options.

dhilton committed:
Merge branch '98091996-deploy-mulitple-apps-tsuru' of https://github.com/alphagov/tsuru-performance-tests into 98091996-deploy-mulitple-apps-tsuru

dhilton committed:
Alignment update to cleanup_apps hash

dhilton committed:
Updated README to document headless Jmeter running

keymon committed:
Add a script to automatically amend the jmx file

Added a script to programatically amend the jmeter.jmx file to
add the `SystemSampler` stanza, due the bug https://github.com/flood-io/ruby-jmeter/issues/43
